File: src/main/java/com/telenordigital/prime/events/ProductDescriptionCacheImpl.java
package com.telenordigital.prime.events;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;
import java.util.TreeMap;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;


public final class ProductDescriptionCacheImpl implements ProductDescriptionCache {

    private static final Logger LOG = LoggerFactory.getLogger(ProductDescriptionCacheImpl.class);

    public static final Product DATA_TOPUP_3GB =
            new Product("DataTopup3GB", new TopUpProduct(3000000000L));

    private static final ProductDescriptionCacheImpl INSTANCE = new ProductDescriptionCacheImpl();

    private final Map<String, Product> products;

    public static ProductDescriptionCacheImpl getInstance() {
        return INSTANCE;
    }


    private ProductDescriptionCacheImpl() {
        products = new TreeMap<>();
        products.put(DATA_TOPUP_3GB.getSku(), DATA_TOPUP_3GB);
    }

    public synchronized void addProduct(final Product p) {
        checkNotNull(p);
        products.put(p.getSku(), p);
    }

    @Override
    public void addTopupProduct(final String sku, final long noOfBytes) {
        checkNotNull(sku);
        checkArgument(noOfBytes >= 0);
        final Product topupProduct = newTopupProduct(sku, noOfBytes);
        LOG.info("Adding topup product  " + topupProduct);
        addProduct(topupProduct);
    }

    private synchronized Product newTopupProduct(final String sku, final long noOfBytes) {
        checkNotNull(sku);
        checkArgument(noOfBytes >= 0);
        return new Product(sku, new TopUpProduct(noOfBytes));
    }

    @Override
    public synchronized boolean isValidSKU(final String sku) {
        checkNotNull(sku);
        return products.containsKey(sku);
    }

    @Override
    public synchronized Product getProductForSku(final String sku) {
        checkNotNull(sku);
        return products.get(sku);
    }
}

File: prime/src/main/kotlin/org/ostelco/prime/storage/ProductDescriptionCacheImpl.kt
package org.ostelco.prime.storage

import com.google.common.base.Preconditions.checkArgument
import com.google.common.base.Preconditions.checkNotNull
import org.ostelco.prime.logger
import org.ostelco.prime.storage.entities.Product
import org.ostelco.prime.storage.entities.TopUpProduct
import java.util.*


object ProductDescriptionCacheImpl : ProductDescriptionCache {

    private val LOG by logger()

    private val products: MutableMap<String, Product>
    val DATA_TOPUP_3GB = Product("DataTopup3GB", TopUpProduct(3000000000L))

    init {
        products = TreeMap()
        products[DATA_TOPUP_3GB.sku] = DATA_TOPUP_3GB
    }

    @Synchronized
    fun addProduct(p: Product) {
        checkNotNull(p)
        products[p.sku] = p
    }

    override fun addTopupProduct(sku: String, noOfBytes: Long) {
        checkNotNull(sku)
        checkArgument(noOfBytes >= 0)
        val topupProduct = newTopupProduct(sku, noOfBytes)
        LOG.info("Adding topup product  " + topupProduct)
        addProduct(topupProduct)
    }

    @Synchronized
    private fun newTopupProduct(sku: String, noOfBytes: Long): Product {
        checkNotNull(sku)
        checkArgument(noOfBytes >= 0)
        return Product(sku, TopUpProduct(noOfBytes))
    }

    @Synchronized
    override fun isValidSKU(sku: String): Boolean {
        checkNotNull(sku)
        return products.containsKey(sku)
    }

    @Synchronized
    override fun getProductForSku(sku: String): Product? {
        checkNotNull(sku)
        return products[sku]
    }
}
